java-checks/src/main/java/org/sonar/java/symexec/ExecutionState.java

```diff
@@ -21,153 +21,107 @@
-    this.relations = HashBasedTable.create();
+    parentState = null;
+    relations = HashBasedTable.create();
   }
 
   ExecutionState(ExecutionState parentState) {
     this.parentState = parentState;
     this.relations = HashBasedTable.create();
+    this.stack = parentState.stack;
   }
 
-  @VisibleForTesting
-  SymbolicRelation getRelation(SymbolicValue leftValue, SymbolicValue rightValue) {
-    SymbolicRelation result = relations.get(leftValue, rightValue);
-    return result != null ? result : parentState != null ? parentState.getRelation(leftValue, rightValue) : UNKNOWN;
+  SymbolicValue peek() {
+    return stack.value;
   }
 
-  SymbolicBooleanConstraint evaluateRelation(SymbolicValue leftValue, SymbolicRelation relation, SymbolicValue rightValue) {
-    return RELATION_RELATION_MAP.get(getRelation(leftValue, rightValue), relation);
+  SymbolicValue pop() {
+    SymbolicValue result = stack.value;
+    stack = stack.parent;
+    return result;
   }
 
-  ExecutionState setRelation(SymbolicValue leftValue, SymbolicRelation relation, SymbolicValue rightValue) {
-    if (!leftValue.equals(rightValue)) {
-      relations.put(leftValue, rightValue, relation);
-      relations.put(rightValue, leftValue, relation.swap());
+  SymbolicValue popLast() {
+    SymbolicValue result = stack.value;
+    stack = stack.parent;
+    if (stack != null) {
+      throw new IllegalStateException("stack is not empty");
     }
+    return result;
+  }
+
+  ExecutionState push(SymbolicValue value) {
+    stack = new StackEntry(stack, value);
     return this;
   }
 
-  void mergeRelations(Iterable<ExecutionState> states) {
-    for (Map.Entry<SymbolicValue, SymbolicValue> entry : findRelatedValues(states).entries()) {
-      SymbolicRelation relation = null;
-      for (ExecutionState state : states) {
-        relation = state.getRelation(entry.getKey(), entry.getValue()).union(relation);
-      }
-      if (relation == null) {
-        relation = SymbolicRelation.UNKNOWN;
-      }
-      if (getRelation(entry.getKey(), entry.getValue()) != relation) {
-        relations.put(entry.getKey(), entry.getValue(), relation);
-        relations.put(entry.getValue(), entry.getKey(), relation.swap());
+  @VisibleForTesting
+  SymbolicRelation getRelation(SymbolicValue leftValue, SymbolicValue rightValue) {
+    for (ExecutionState executionState = this; executionState != null; executionState = executionState.parentState) {
+      SymbolicRelation result = executionState.relations.get(leftValue, rightValue);
+      if (result != null) {
+        return result;
       }
     }
+    return UNKNOWN;
+  }
+
+  SymbolicValue evaluateRelation(SymbolicValue leftValue, SymbolicRelation relation, SymbolicValue rightValue) {
+    switch (getRelation(leftValue, rightValue).combine(relation)) {
+      case FALSE:
+        return SymbolicValue.BOOLEAN_FALSE;
+      case TRUE:
+        return SymbolicValue.BOOLEAN_TRUE;
+      default:
+        return new SymbolicValue();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this instantiation of "SymbolicValue". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)